app/src/main/java/de/westnordost/streetcomplete/measure/MeasureActivity.kt

```diff
@@ -0,0 +1,418 @@
+        }
+    }
+
+    private fun measuringDone() {
+        binding.arSceneViewContainer.performHapticFeedback(VIRTUAL_KEY)
+        if (requestResult) binding.acceptResultContainer.isGone = false
+        measureState = MeasureState.DONE
+    }
+
+    private fun clearMeasuring() {
+        measureState = MeasureState.READY
+        binding.arSceneViewContainer.performHapticFeedback(VIRTUAL_KEY)
+        binding.measurementSpeechBubble.isInvisible = true
+        binding.acceptResultContainer.isGone = true
+        distance = 0f
+        firstNode?.anchor?.detach()
+        firstNode?.setParent(null)
+        firstNode = null
+        secondNode?.anchor?.detach()
+        secondNode?.setParent(null)
+        secondNode = null
+        lineNode?.setParent(null)
+        lineNode = null
+    }
+
+    private fun returnMeasuringResult() {
+        val resultIntent = Intent(RESULT_ACTION)
+        resultIntent.putExtra(RESULT_MEASURE_IN_METERS, distance)
+        setResult(RESULT_OK, resultIntent)
+        finish()
+    }
+
+    private fun updateCursor(hitResult: HitResult) {
+        // release previous anchor only if it is not used by any other node
+        val anchor = cursorNode?.anchor
+        if (anchor != null && anchor != firstNode?.anchor && anchor != secondNode?.anchor) {
+            anchor.detach()
+        }
+
+        try {
+            val newAnchor = hitResult.createAnchor()
+            val cursorNode = getCursorNode()
+            cursorNode.anchor = newAnchor
+
+            if (measureState == MeasureState.MEASURING) {
+                secondNode?.anchor = newAnchor
+            }
+            /* update distance should always be called because the world position could be adjusted
+             *  for already existing anchors */
+            updateDistance()
+        } catch (e: Exception) {
+            Log.e("MeasureActivity", "Error", e)
+        }
+    }
+
+    private fun updateDistance() {
+        val pos1 = firstNode?.worldPosition
+        val pos2 = secondNode?.worldPosition
+        val up = firstNode?.up
+        val hasMeasurement = pos1 != null && pos2 != null && up != null
+
+        binding.measurementSpeechBubble.isInvisible = !hasMeasurement
+        if (!hasMeasurement) return
+
+        val difference = Vector3.subtract(pos1, pos2)
+        distance = difference.length()
+
+        binding.measurementTextView.text = displayUnit.format(distance.toDouble())
+
+        val line = getLineNode()
+        line.worldPosition = Vector3.add(pos1, pos2).scaled(.5f)
+        line.worldRotation = Quaternion.lookRotation(difference, up)
+        line.localScale = Vector3(1f, 1f, distance)
+    }
+
+    private fun getCursorNode(): AnchorNode {
+        var node = cursorNode
+        if (node == null)  {
+            node = AnchorNode().apply {
+                renderable = cursorRenderable
+                setParent(arSceneView!!.scene)
+            }
+            cursorNode = node
+        }
+        return node
```

Comment:
I think it should be equivalent and a bit more readable without the local variable:

```suggestion
        if (cursorNode == null)  {
            cursorNode = AnchorNode().apply {
                renderable = cursorRenderable
                setParent(arSceneView!!.scene)
            }
        }
        return cursorNode
```

Reply:
`cursorNode` is a nullable field, so this is not equivalent (function returns non-nullable)